### PR TITLE
Add timer to trigger periodic sync every 120s

### PR DESCRIPTION
**Pull request by dtzWill**
Only lightly tested for now, hopefully works for others too :).

cc #74.

### CODE REVIEW

File: CMakeLists.txt
Comment (ohwgiles) [inline CMakeLists.txt]: This doesn't look relevant. If it was necessary, we should break it into a separate commit with an explanation why

File: src/main.c
Comment (ohwgiles) [inline src/main.c]: No need to involve gdk_threads. Suggest [g_timeout_add_seconds](https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#g-timeout-add-seconds)

**Review by ohwgiles — CHANGES_REQUESTED** [review]
Hi, thanks for contributing! I'm actually working on this issue at the moment, there are some tricky bits like handling a remote deletion while you're editing an event. Anyway I hadn't got to actually triggering the timer yet, so I'd be happy to merge this in if you address a few minor things :)

File: CMakeLists.txt
Comment (dtzWill) [reply CMakeLists.txt]: Sure thing, agreed it should be split out-- I was touching these files while working on this, actually :innocent:.  First pass used glib, so was detecting that and plumbing flags accordingly.

Anyway this is a minor fixup, split off into #80 .

File: src/main.c
Comment (dtzWill) [reply src/main.c]: Okay, no problem -- I wasn't sure which approach to take and am new to gtk/glib programming :).  I actually used g_timeout_add originally I think, will switch back.  I don't remember what motivated switching-- I think it was in response to reading documentation text like this:

> You should use gdk_threads_add_idle() and gdk_threads_add_timeout() instead of g_idle_add() and g_timeout_add() since libraries not under your control might be using the deprecated GDK locking mechanism. If you are sure that none of the code in your application and libraries use the deprecated gdk_threads_enter() or gdk_threads_leave() methods, then you can safely use g_idle_add() and g_timeout_add().
(from https://developer.gnome.org/gdk3/stable/gdk3-Threads.html)

I have not worked with these details before, and reading bits like the above and other comments led me to believe it was "safer" to use gdk_threads to ensure appropriate locks are held (perhaps facilitating issues due to racing with other activities?).  Anyway I happily defer to almost anyone's suggestions, having little experience with this myself :).

Looking into things a bit further turns up instances such as[1] which sure makes it sound like your suggested approach is the way to go -- and gdk lock I thought was useful and safe is instead slated for removal? Hah, whoops!

Will update in a few, I believe I have the other approach on hand.

[1] https://gitlab.gnome.org/GNOME/gtk/commit/c655759cefa56efdb0cb943bb5dada3ff75a963c

Comment (dtzWill) [reply src/main.c]: Oh, yes. I think I wasn't sure how to cleanly ensure the callback didn't just leak and only fired with a valid FocalApp pointer.

https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#mainloop-memory-management is what gave me some doubts.

(Apologies, as you can perhaps tell I have not worked with any GTK/glib applications before; much of what I know came from documentation and looking at examples in the gtk tree -- including a calendar sync one actually! haha)

Anyway I'll see what I can do.  If nothing else, maybe the primary benefit of this PR will be to let you know someone's interested and provide a smidge of motivation :grin: .  Ahem, one moment...

**Comment by dtzWill** [comment]
I kept the pattern of retaining the source ID and freeing as part of shutdown.  I'm not quite clear how everything fits together, so hopefully this makes sense :innocent:.

File: src/main.c
Comment (ohwgiles) [reply src/main.c]: > https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#mainloop-memory-management is what gave me some doubts.

The "first, and preferred" option mentioned there is exactly what you do in this PR - great.

Generally, focal is an asynchronous *single-threaded* application, so there should be no need to worry about threads at all

File: CMakeLists.txt
Comment (ohwgiles) [inline CMakeLists.txt]: glib is an intrinsic dependency of gtk, so you don't need to add it specifically. In fact, if you search the focal code, you should find extant instances of `g_timeout_add_seconds` already.

File: CMakeLists.txt
Comment (ohwgiles) [inline CMakeLists.txt]: as above

File: src/main.c
Comment (ohwgiles) [inline src/main.c]: Not needed

File: src/main.c
Comment (ohwgiles) [inline src/main.c]: GObject-based objects should be cast with their helper macros, i.e. `FOCAL_APP(userdata)`. This allows a runtime check that the object is indeed a `FocalApp`. There are plenty of examples of this throughout this file

File: src/main.c
Comment (ohwgiles) [inline src/main.c]: Comment is a little redundant thanks to the `_seconds` in the API. Then it can easily fit on one line and save vertical space. Not critical, just since we're discussing this...

File: src/main.c
Comment (ohwgiles) [inline src/main.c:425]: :+1:

File: src/main.c
Comment (ohwgiles) [inline src/main.c:369]: Just curious because I haven't used this in other `g_timeout_add` calls, what is this useful for?

File: CMakeLists.txt
Comment (dtzWill) [reply CMakeLists.txt]: Hmm, I can't think of any reason I'd add the dependency if i didn't run into a problem.  But indeed the code already uses this API, and it doesn't seem to be required.  I *did* have fun indulging in learning and such-- no apologies there but I am sorry for not better cleaning this all up before submitting this PR! Thanks for working with me on this :).

Pretty sure this PR will come down to basically 3 semi-trivial lines! Oh well, less lines of code means less code to maintain ;).

File: src/main.c
Comment (dtzWill) [reply src/main.c]: Noted, and neat! Thanks, will update accordingly!

File: src/main.c
Comment (dtzWill) [reply src/main.c]: Agreed, thanks for your review O:).

File: src/main.c
Comment (dtzWill) [reply src/main.c:369]: Not entirely sure, but example I was looking at used it and it seemed potentially useful.  Maybe not-- I was still piecing together what a source was, and so I was hardly in a position to snub some debuggability when it was right there :).  I did, however, imagine it might end up as the thread name, but that doesn't appear to be the case.  Which is good because spawning polling thread for each source does seem a bit silly.  Anyway, since you asked I and because I was curious myself, here's what the docs[1] say:

> Sets the name of a source using its ID.
> 
> This is a convenience utility to set source names from the return value of g_idle_add(), g_timeout_add(), etc.
> 
> It is a programmer error to attempt to set the name of a non-existent source.
> 
> More specifically: source IDs can be reissued after a source has been destroyed and therefore it is never valid to use this function with a source ID which may have already been removed. An example is when scheduling an idle to run in another thread with g_idle_add(): the idle may already have run and been removed by the time this function is called on its (now invalid) source ID. This source ID may have been reissued, leading to the operation being performed against the wrong source.
> 
> 

`g_source_set_name` adds:

> 
> Sets a name for the source, used in debugging and profiling. The name defaults to NULL.
> 
> The source name should describe in a human-readable way what the source does. For example, "X11 event queue" or "GTK+ repaint idle handler" or whatever it is.
> 
> It is permitted to call this function multiple times, but is not recommended due to the potential performance impact. For example, one could change the name in the "check" function of a GSourceFuncs to include details like the event type in the source name.
> 
> Use caution if changing the name while another thread may be accessing it with g_source_get_name(); that function does not copy the value, and changing the value will free it while the other thread may be attempting to use it.

I'm still slightly unclear what, concretely, might access and display this name.  Maybe some of glib/gtk logging or assertion macros use this as a bit of execution context info?

I am happy with whatever your preference is here, maybe later I'll chase down a better answer as to who might see this name, or said another way: "why is this not essentially a comment with runtime cost" :laughing: 

[1] https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#g-source-set-name-by-id

File: src/main.c
Comment (dtzWill) [reply src/main.c:425]: Since I checked at the time, FWIW unlike `free()` and such, the conditional is required as a valid source must be passed and zero is not valid.  Or so I recall. :).

File: CMakeLists.txt
Comment (ohwgiles) [reply CMakeLists.txt]: > Pretty sure this PR will come down to basically 3 semi-trivial lines

This is of course always the best result

**Review by ohwgiles — APPROVED** [review]

**Comment by ohwgiles** [comment]
Looks ready to merge. I'll hold off until the weekend because I want it to go in with my other changes related to automatic sync and I should have time to tidy the last bits up then.

**Comment by dtzWill** [comment]
On Thu, 24 Oct 2019 10:52:54 -0700, Oliver Giles <notifications@github.com> wrote:
> Looks ready to merge. I'll hold off until the weekend because I want it to go in with my other changes related to automatic sync and I should have time to tidy the last bits up then.
> 

:+1: -- sounds great, thanks!

**Pull request merged**
